[Zhao Wenrui] iP

wenruu opened the pull request:
# Botato
> What I say is that, if a fellow really likes potatoes, he must be a pretty decent sort of fellow. </br>- A. A. Milne [(source)](https://www.goodreads.com/quotes/8447-what-i-say-is-that-if-a-man-really-likes#:~:text=Sign%20Up%20Now-,What%20I%20say%20is%20that%2C%20if%20a%20man%20really%20likes,pretty%20decent%20sort%20of%20fellow.)

Botato is the smartest potato 🥔 (who's a bot) you'll come across! It's,
- Intuitive
- Helpful
- Convenient

All you need to do is,
1. Download it from [here](https://github.com/wenruu/ip/releases/).
2. Run it.
3. Add your tasks.
4. Let it manage your tasks!

Best of all, it's **FREE!**

Features:
- [x] Create and remember your tasks
- [x] Manage deadlines
- [ ] AI integration (coming soon!)

---
This is an example of the main method used:
```java
public static void main(String[] args) {
    new Botato().run();  // Create an instance of Botato and run the application.
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

bl3hz committed:
Level-0

bl3hz committed:
Level-0

bl3hz committed:
Level-1

bl3hz committed:
Level-2

bl3hz committed:
Level-3

bl3hz committed:
Level-3.1

bl3hz committed:
Level-3.2

wenruu committed:
Level-4

wenruu committed:
A-TextUiTesting

wenruu committed:
Level-5

wenruu committed:
Level-6

wenruu committed:
Level-6.1 (Bugfixes and made some code more concise). A-Enum not needed as classes inherit from Task.

wenruu committed:
branch-Level-7

wenruu committed:
Merge branch 'branch-Level-7'

wenruu committed:
branch-Level-8

wenruu committed:
Merge branch 'branch-Level-8'

wenruu committed:
Add classes for OOP and sorted into packages.

Much of the previous code needs to be refactored to achieve OOP. Relevant classes need to be made to facilitate this.

Transferring code to these classes creates a more lightweight main class and improves code quality.

Let's,
	- Keep transferring code from the main class
	- Refactor all the code to an OOP context

bl3hz committed:
Merge branch 'master' of https://github.com/wenruu/ip

bl3hz committed:
Add more code and bugfixes to OOP classes.

Let's,
- Continue fixing bugs and adding code
- Test and see if OOP code performs as expected

bl3hz committed:
Add code to OOP classes and fix bugs.

Let's,
- Keep adding more code and fixing bugs

wenruu committed:
Finish refactoring code to OOP standards.

wenruu committed:
Merge branch 'A-MoreOOP'

* A-MoreOOP:
  Finish refactoring code to OOP standards.
  Add code to OOP classes and fix bugs.
  Add more code and bugfixes to OOP classes.
  Add classes for OOP and sorted into packages.

wenruu committed:
Add packages.

wenruu committed:
Merge branch 'A-Packages'

* A-Packages:
  Add packages.

wenruu committed:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

* commit '81a9c537afe8492675eb253968939364d65db091':
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

wenruu committed:
Add gradle to project.

AnWe11 commented on src/main/java/botato/Botato.java:
```diff
@@ -0,0 +1,42 @@
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
+        while (!isExit) {
+            try {
+                String cmd = ui.readCommand();
+                ui.showLine();
+                Command c = Parser.parse(cmd);
+                c.execute(tasks, ui);
+                isExit = c.isExit();
+            } catch (BotatoException e) {
+                ui.showError(e.getMessage());
+            } finally {
+                ui.showLine();
+            }
+        }
```
You could also use a do{} while() loop and only exit when the command is "bye".

AnWe11 commented on src/main/java/botato/Parser.java:
```diff
@@ -0,0 +1,46 @@
+
+public class Parser {
+    public static Command parse(String cmd) {
+        if (cmd.equals("bye")) {
+            // Save data and exit chatbot
+            return new ExitCommand();
+        } else if (cmd.equals("list")) {
+            // Display list of tasks
+            return new ListCommand();
+        } else if (cmd.startsWith("mark ")) {
+            // Mark specific task as completed
+            return new MarkCommand(cmd);
+        } else if (cmd.startsWith("unmark ")) {
+            // Mark specified task as not done
+                return new UnmarkCommand(cmd);
+        } else if (cmd.startsWith("todo ")){
+            // Add a Task of type Todo to tasks
+            return new TodoCommand(cmd);
+        } else if (cmd.startsWith("deadline")){
+            // Add a Task of type Deadline to tasks
+            return new DeadlineCommand(cmd);
+        } else if (cmd.startsWith("event ")) {
+            // Add a Task of type Event to tasks
+            return new EventCommand(cmd);
+        } else if (cmd.startsWith("delete ")) {
+            // Delete a specified task based on task number given
+            return new DeleteCommand(cmd);
+        } else if (cmd.equals("help")) {
+            // Prints out available commands
+            return new HelpCommand();
+        } else {
+            // Handle unknown commands
+            throw new UnknownCommandException();
+        }
```
Maybe a switch statement would be better?

AnWe11 commented on src/main/java/botato/Storage.java:
```diff
@@ -0,0 +1,37 @@
+        try (FileInputStream fileIn = new FileInputStream("./data/data.txt");
+             ObjectInputStream in = new ObjectInputStream(fileIn)) {
+            Object obj = in.readObject();
+            @SuppressWarnings("unchecked")
```
Not sure if it is good to use @SuppressWarnings

AnWe11 commented on src/main/java/botato/Storage.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.io.*;
+import java.util.ArrayList;
+
```
You are missing your descriptive header which summarises what the class does

AnWe11 commented on src/main/java/botato/task/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Represents a Deadline task, which is a specific type of Task.
+ * A Deadline task has a description and a deadline (stored in the {@code by} field).
+ * This class extends the {@link Task} class and overrides the {@code toString()}
+ * method to provide a custom string representation of the task, including the deadline.
+ * <p>
+ * The description and deadline are extracted from the input command string.
+ * The description is the part of the command before the "/by" keyword, and the deadline
+ * is the part after the "/by" keyword.
+ * <p>
+ * Example usage:
+ * <pre>
+ * Deadline deadline = new Deadline("deadline Submit report /by 2023-10-31");
+ * System.out.println(deadline); // Outputs: "[ ] Deadline: Submit report (by: 2023-10-31)"
+ * </pre>
+ */
```
Description too long, although I understand the explanation, you should keep the description succinct and summarise the class. I dont think examples would be necessary.

AnWe11 commented on src/main/java/botato/task/Task.java:
```diff
@@ -0,0 +1,112 @@
+                LocalDate date = LocalDate.parse(dateStr);
+                return LocalDateTime.of(date, defaultTime);
+            }
+            if (dateStr.matches("\\d{2}/\\d{2}/\\d{4} \\d{2}:\\d{2}:\\d{2}")) {
+                return LocalDateTime.parse(dateStr, DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss"));
+            }
+            if (dateStr.matches("\\d{2}/\\d{2}/\\d{4} \\d{2}:\\d{2}")) {
+                return LocalDateTime.parse(dateStr, DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm"));
+            }
+            if (dateStr.matches("\\d{2}/\\d{2}/\\d{4}")) {
+                LocalDate date = LocalDate.parse(dateStr, DateTimeFormatter.ofPattern("dd/MM/yyyy"));
+                return LocalDateTime.of(date, defaultTime);
+            }
+            if (dateStr.matches("\\d{2}/\\d{2}/\\d{2} \\d{2}:\\d{2}:\\d{2}")) {
+                return LocalDateTime.parse(dateStr, DateTimeFormatter.ofPattern("dd/MM/yy HH:mm:ss"));
+            }
+            if (dateStr.matches("\\d{2}/\\d{2}/\\d{2} \\d{2}:\\d{2}")) {
+                return LocalDateTime.parse(dateStr, DateTimeFormatter.ofPattern("dd/MM/yy HH:mm"));
+            }
+            if (dateStr.matches("\\d{2}/\\d{2}/\\d{2}")) {
+                LocalDate date = LocalDate.parse(dateStr, DateTimeFormatter.ofPattern("dd/MM/yy"));
+                return LocalDateTime.of(date, defaultTime);
+            }
+            if (dateStr.matches("\\d{2} \\w{3} \\d{4} \\d{2}:\\d{2}:\\d{2}")) {
+                return LocalDateTime.parse(dateStr, DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm:ss"));
+            }
+            if (dateStr.matches("\\d{2} \\w{3} \\d{4} \\d{2}:\\d{2}")) {
+                return LocalDateTime.parse(dateStr, DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm"));
+            }
+            if (dateStr.matches("\\d{2} \\w{3} \\d{4}")) {
+                LocalDate date = LocalDate.parse(dateStr, DateTimeFormatter.ofPattern("dd MMM yyyy"));
+                return LocalDateTime.of(date, defaultTime);
+            }
+        } catch (DateTimeParseException e) {
+            throw new InvalidDateTimeFormatException();
+        }
+        return null;
+    }
+
+}
```
Maybe this would be better as a class on its own?

AnWe11 reviewed (commented):
Overall, code is easy to follow but I feel that a little more abstraction could be done, regarding the timeParser in the Task class. Class descriptions were also a bit too long, should try to keep them succinct and summarised. Some functions that were not getters or setters also did not have documentations. Naming conventions were consistent and correct. Overall, good job!

wenruu committed:
Add JUnit tests for a few classes.

wenruu committed:
Add coding standards (already adhered to) and Branch-Level-9

Mistakenly wrote code on wrong branch so will commit here instead of copying over.

wenruu committed:
Add find command

Allows user to find a task in their task list, enhancing user experience.

wenruu committed:
Add JavaDocs for Parser.java, Storage.java, TaskList.java

JavaDocs was not written yet.

Adding JavaDocs makes it more clear what the classes do.

Let's,
- Keep adding JavaDocs for the remaining non-private classes.

bl3hz committed:
Add JavaDocs to more classes

bl3hz committed:
Revert "Add JavaDocs to more classes"

This reverts commit 2c8cde71f967831991e5b66891e626311ece32d6.

bl3hz committed:
Revert "Revert "Add JavaDocs to more classes""

This reverts commit bafa55f61def55d77222cc2e15cf6e9f23a94072.

bl3hz committed:
Merge branch 'branch-Level-9'

bl3hz committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/botato/TaskList.java

bl3hz committed:
Fix merge conflict

wenruu committed:
Merge branch 'branch-A-JavaDoc' of github.com:wenruu/ip into branch-A-JavaDoc

* 'branch-A-JavaDoc' of github.com:wenruu/ip:
  Add JavaDocs to more classes

wenruu committed:
Merge branch 'branch-A-JavaDoc'

wenruu committed:
feat: code style refactoring

Code previously was not adherent to the code style guidelines of CS2103T.

This has been fixed using CheckStyle.

wenruu committed:
Integrate GUI to Botato

Botato was only text-based, making the user experience less than ideal.

Adding a GUI makes using the chatbot more intuitive and aesthetically
pleasing.

Integrating JavaFX, and using FXML files to define the user interface
along with CSS stylesheets allows us to create a beautiful, easily
editable GUI.

Integrating JavaFX allows us to create a GUI. Using FXML lets us use
SceneBuilder to edit the GUI layout graphically instead of
programatically. Adding CSS stylesheets lets us easily style and edit
the design of the GUI. For example, with SceneBuilder, we can tweak the
design and immediately see the changes without having to run the
application.

Let's further customise the design by,
* Adding a custom background to the chat
* Tweaking the designs and colours of the application
* Add more functionalities to the application

wenruu committed:
Add image resources missing in last commit

wenruu committed:
Merge branch 'Level-10'

wenruu committed:
Refactor parser, add asserts

Validating task numbers is handled in the commands that use them.

This causes unnecessary code duplication and reduces code clarity.

Let's validate task numbers in the parser instead.

This makes behaviour clearer and reduces code duplication.

Furthermore, assertions were added to commands using task numbers to
ensure task numbers are valid and prevent null pointer errors.

wenruu committed:
Merge pull request #2 from wenruu/A-Assertions

Refactor parser, add asserts

wenruu committed:
Merge branch 'master' of github.com:wenruu/ip

wenruu committed:
Improve code quality

wenruu committed:
Change GUI design

Added a custom background and changed several elements including font
size, colour scheme, and profile image of Botato.

wenruu committed:
Merge pull request #3 from wenruu/A-CodeQuality

A-CodeQuality

wenruu committed:
Decrease transparency of background

Background was clashing with other UI design elements, making it hard
to see the dialog boxes and profile pictures.

This negatively impacts the user experience.

Reducing the background image transparency to 30 makes other UI elements
much clearer while keeping the image discernible compared to blurring.

In addition, background image was shifted to the correct directory in
resources, under "images" now.

wenruu committed:
Minor bugfix

User can set their end date to be before the start date when creating
"Event" tasks.

This obviously defies logic and should be changed.

Creating a new exception to be thrown, and throwing it when the end date
is before the start date, fixes the problem.

This ensure that the exception can be caught as per normal, and the
message should inform the user that they might need to get their eyes
(or brain) checked.

wenruu committed:
Add Ui.png

wenruu committed:
Update user guide

wenruu committed:
User guide fix

wenruu committed:
Add _config.yml

wenruu committed:
Add custom background to user guide

wenruu committed:
Update Jekyll theme

wenruu committed:
Change background image

wenruu committed:
Update background

wenruu committed:
Change user guide background

wenruu committed:
Change user guide background

wenruu committed:
Change user guide background

wenruu committed:
Adjust font and dialog box colours

Design for both user and botato dialog boxes is exactly the same, with a
gradient.

This makes it often confusing who's who and the gradient can make words
harder to discern sometimes.

Changing each the user's and botato's dialog boxes to different
monotonous colours makes who's saying what clearer. Changing botato's
font colour to a lighter one makes it easier to read.

Botato was given the lighter shade because it feels more intuitive that
the user's dialog box is darker than the person replying. Making
colours monotonous allows us to optimise text to give a higher contrast,
making both dialog boxes more readable.

wenruu committed:
Improve help functionality

Help functionality is purely text-based and unintuitive, with no
examples shown to the user.

This makes the help not very helpful, and users may have to keep
scrolling up and down to copy and find the information they need.

Let's make the app open a new window containing the user guide when
the user asks for help.

As we already have a polished user guide (README.md), it would be
easier and better to just reuse that for the internal help function,
as updating the guide as more functionalities are added is much easier
this way. In addition, the user may be using the app offline, so just
opening the app's website is not good.

A new dependency, FlexMark, was added to convert .md to HTML (as JavaFX
does not natively support reading .md files), and the HTML file is
opened in a new WebView using JavaFX.

In addition, updated the parser to read user inputs for single-word
commands (help, list etc) even if there are leading or trailing
whitespaces, in the spirit of fat fingers.

Let's,
* Make sure the new dependency is properly bundled into the .jar file
  when running shadowJar
* Test that the user guide opens correctly in other systems

wenruu committed:
Update JUnit tests

wenruu committed:
Fix user guide missing in fat JAR

wenruu committed:
Add missing dependencies
